### PR TITLE
PROTOTECH-42: Rounding In Favor of the Interacting User Is Dangerous

### PR DESCRIPTION
**Pull request by mattcushman**
# Description of change
## High level
* See https://github.com/Fixed-Point-Solutions/prototech-ajna-audit/issues/42
  * Unify the way LP redeemed and awarded are calculated and ensure they're using proper direction when rounded (up when LP redeemed, down when LP awarded). This solves the big deviations of exchange rate noticed in invariant tests conducted by Prototech team
  * Included Prototech regression tests related to exchange rate invariant (R1 - R5, R8, R9, R10, R12, R15 - R19)

# Description of bug or vulnerability and solution
* Rounding In Favor of the Interacting User Is Dangerous and could lead to protocol loses
* introduced Buckets library assets to LP and LP to assets conversion helpers:
  - used in all contracts where LP are calculated (instead duplicated logic and calculate LP using `Buckets.exchangeRate`)
  - use OZ library `Math.mulDiv` to perform conversion when bucket has deposit, collateral and LP
  - accept direction of rounding (up or down) as parameter

* Assets to LP helpers:
  * `collateralToLP`:
    - returns:
amount of LP calculated as `collateral * bucket price` if there's no deposit, collateral or LP in bucket, rounded nearest
otherwise `bucket LP * (collateral to convert * bucket price) / (deposit + bucket collateral * bucket price)` and rounds accordingly
    - used in:
`LenderActions.removeCollateral` (ERC721Pool) and `LenderActions._removeMaxCollateral` (ERC20Pool) to calculate amount of LP redemeed for removed collateral, rounded up
`Buckets.addCollateral` to calculate amount of LP awarded for the amount of collateral, rounded down

  * `quoteTokensToLP`:
    - returns:
amount of LP equal with amount of quote token to convert if there's no deposit, collateral or LP in bucket
otherwise `(bucket LP * quote token amount) / (deposit + bucket collateral * bucket price)` and rounds accordingly
    - used in:
`KickerActions.kickWithDeposit` to calculate amount of LP redeemed for quote tokens used to kick auction, rounded up
`LenderActions.addQuoteToken` to calculate amount of LP awarded for the amount of quote tokens added, rounded down
`LenderActions._removeMaxDeposit` (used in add and move quote token) to calculate amount of LP redeemed for the amount of quote tokens removed from bucket, rounded up
`LenderActions.moveQuoteToken` to calculate amount of LP awarded for the amount of quote tokens moved in to bucket, rounded down
`TakerActions._rewardBucketTake` to calculate taker and kicker LP rewards, rounded down

* LP to assets helpers:
  * `lpToCollateral`:
    - returns:
amount of collateral calculated as `LP * bucket price` if there's no deposit, collateral or LP in bucket, rounded nearest
otherwise `(deposit + bucket collateral * bucket price) * LP / (bucket LP * bucket price)` and rounds accordingly
    - does not cap amount at collateral available in bucket
    - used in `PoolHelper._lpToCollateral` to calculate amount of available collateral in bucket for given LP amount, rounded down

  * `lpToQuoteTokens`:
    - returns:
amount of quote tokens equal with amount of LP to convert if there's no deposit, collateral or LP in bucket
otherwise `(deposit + bucket collateral * bucket price) * LP / bucket LP` and rounds accordingly
    - does not cap amount at quote tokens available in bucket
    - used in:
`KickerActions.kickWithDeposit` to calculate amount of quote token to debit from deposit when kick with deposit, rounded down
`PoolHelper._lpToQuoteToken` to calculate amount of available quote tokens in bucket for given LP amount, rounded down

# Contract size
## Pre Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,470B  (99.56%)
  ERC20Pool                -  23,819B  (96.92%)
  PositionManager          -  19,300B  (78.53%)
  TakerActions             -  14,949B  (60.83%)
  BorrowerActions          -  12,215B  (49.70%)
  LenderActions            -  10,023B  (40.78%)
  KickerActions            -   9,685B  (39.41%)
  PoolCommons              -   9,233B  (37.57%)
  SettlerActions           -   9,189B  (37.39%)
```
## Post Change
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,470B  (99.56%)
  ERC20Pool                -  23,819B  (96.92%)
  PositionManager          -  19,532B  (79.47%)
  TakerActions             -  15,228B  (61.96%)
  BorrowerActions          -  12,420B  (50.54%)
  LenderActions            -  10,496B  (42.71%)
  KickerActions            -  10,113B  (41.15%)
  SettlerActions           -   9,383B  (38.18%)
  PoolCommons              -   9,233B  (37.57%)
```

# Gas usage
## Pre Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120727          | 177942 | 162586 | 712465  | 60004   |
| bucketTake                           | 308615          | 326982 | 326982 | 345350  | 4       |
| drawDebt                             | 250455          | 295356 | 269279 | 809196  | 136003  |
| kick                                 | 147490          | 221511 | 220012 | 1032365 | 48000   |
| kickWithDeposit                      | 196696          | 274344 | 270489 | 996938  | 8000    |
| moveQuoteToken                       | 278302          | 278302 | 278302 | 278302  | 1       |
| removeQuoteToken                     | 144953          | 167579 | 171672 | 195094  | 4000    |
| repayDebt                            | 574996          | 607152 | 597044 | 755414  | 16002   |
| settle                               | 349355          | 349355 | 349355 | 349355  | 1       |
| take                                 | 81142           | 82544  | 82236  | 313980  | 15998   |
```
## Post Change
```
| Function Name                        | min             | avg    | median | max     | # calls |
| addQuoteToken                        | 120387          | 177597 | 162246 | 712049  | 60004   |
| bucketTake                           | 308310          | 326951 | 326951 | 345592  | 4       |
| drawDebt                             | 250455          | 295356 | 269279 | 809196  | 136003  |
| kick                                 | 147494          | 221515 | 220016 | 1032369 | 48000   |
| kickWithDeposit                      | 197204          | 274851 | 270997 | 997366  | 8000    |
| moveQuoteToken                       | 268989          | 268989 | 268989 | 268989  | 1       |
| removeQuoteToken                     | 144953          | 169074 | 173484 | 195974  | 4000    |
| repayDebt                            | 574996          | 607152 | 597044 | 755414  | 16002   |
| settle                               | 349355          | 349355 | 349355 | 349355  | 1       |
| take                                 | 81142           | 82544  | 82236  | 313980  | 15998   |
```

### CODE REVIEW

File: src/libraries/internal/Buckets.sol [hunk 1]
```diff
@@ -98,18 +100,27 @@ library Buckets {
...
         uint256 collateral_,
-        uint256 bucketPrice_
+        uint256 bucketPrice_,
+        Math.Rounding rounding_
     ) internal pure returns (uint256 lp_) {
-        uint256 rate = getExchangeRate(bucketCollateral_, bucketLP_, deposit_, bucketPrice_);
-
-        lp_ = Maths.wdiv(Maths.wmul(collateral_, bucketPrice_), rate);
+        if (deposit_ == 0 && bucketCollateral_ == 0) {
+            lp_ = Maths.wmul(collateral_, bucketPrice_);
```
Comment (kmbarry1) [inline src/libraries/internal/Buckets.sol]: This can go randomly for or against the user still since `Maths.wmul` rounds to the nearest `WAD` (i.e. the `rounding_` argument is not respected here).

File: src/libraries/external/LenderActions.sol [hunk 2]
```diff
@@ -666,8 +673,8 @@ library LenderActions {
             lpAmount_ = requiredLP;
         } else {
             lpAmount_         = lenderLpBalance;
-            collateralAmount_ = Maths.wdiv(Maths.wmul(lenderLpBalance, collateralAmount_), requiredLP);
-
+            collateralAmount_ = Math.mulDiv(lenderLpBalance, collateralAmount_, requiredLP);
+
```
Comment (mattcushman) [inline src/libraries/external/LenderActions.sol]: nit: whitespace

**Review by mattcushman — COMMENTED** [review]
lgtm (left one minor nit comment on ws)

**Comment by prat-gpt** [comment]
Some places where I think the calculation can be updated to favour the pool.

https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/LenderActions.sol#L302 - floorDiv 
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/LenderActions.sol#L190 - floorDiv 
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/BorrowerActions.sol#L316 - ceilMul
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/BorrowerActions.sol#L197 - ceilDiv
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/BorrowerActions.sol#L310- floorDiv 
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/KickerActions.sol#L301 - floorMul
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/KickerActions.sol#L229- ceilMul 
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/SettlerActions.sol#L362 - ceilDiv
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/SettlerActions.sol#L369 - ceilDiv
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/SettlerActions.sol#L135 - floorMul 
https://github.com/ajna-finance/contracts/blob/prototech-R9-fix/src/libraries/external/TakerActions.sol#L296 - ceilMul

**Review by prat-gpt — APPROVED** [review]
LGTM! working on a PR on top of this PR here #851 with some additional rounding updates based on my comment [here](https://github.com/ajna-finance/contracts/pull/838#issuecomment-1559324851).

File: src/libraries/external/KickerActions.sol [hunk 3]
```diff
@@ -223,14 +223,22 @@ library KickerActions {
...
+                Math.Rounding.Up
+            );
 
             uint256 unscaledAmountToRemove = Maths.wdiv(vars.amountToDebitFromDeposit, vars.bucketScale);
 
             // revert if calculated unscaled amount is 0
             if (unscaledAmountToRemove == 0) revert InsufficientLiquidity();
 
             Deposits.unscaledRemove(deposits_, index_, unscaledAmountToRemove);
+
```
Comment (EdNoepel) [inline src/libraries/external/KickerActions.sol]: nit: unnecessary whitespace; two lines are closely related

File: tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol [hunk 4]
```diff
@@ -986,20 +1013,20 @@ contract ERC20PoolCollateralTest is ERC20HelperContract {
             from:    _lender,
             amount:  883976901103343226.563974622543668416 * 1e18,
             index:   2570,
-            lpAward: 2408878317819033617340.926215832879088040 * 1e18
+            lpAward: 2408878346532910443679.386064220627467464 * 1e18
```
Comment (EdNoepel) [inline tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol]: nit: I know "it was like that when I got here", but thousands separators are helpful.

File: tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol [hunk 5]
```diff
@@ -1088,71 +1115,82 @@ contract ERC20PoolCollateralTest is ERC20HelperContract {
...
             exchangeRate: 1 * 1e18           // exchange rate didn't change
         });
 
-        // attacker still has their single LP share
+        // attacker doesn't have any LP share
         _assertLenderLpBalance({
             lender:      attacker,                 // attacker address
             index:       6502,                     // bucket index
-            lpBalance:   1,                        // attacker's LP remaining
+            lpBalance:   0,                        // no LP remaining
```
Comment (EdNoepel) [inline tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol:1193]: nice

File: tests/forge/utils/DSTestPlus.sol [hunk 6]
```diff
@@ -118,8 +119,21 @@ abstract contract DSTestPlus is Test, IPoolEvents {
...
+        ) = _pool.bucketInfo(index);
+
+        uint256 lpAmount = Buckets.quoteTokensToLP(
+            bucketCollateral,
+            bucketLP,
+            bucketDeposit,
+            amount,
+            _priceAt(index),
+            Math.Rounding.Down
+        );
```
Comment (EdNoepel) [inline tests/forge/utils/DSTestPlus.sol]: I don't understand this change.  For initial liquidity, `lpAmount` should always equal the amount of quote token (for 18-decimal quote token).  And this seems to eliminate the scaling operation performed on (old) line 122.

**Review by EdNoepel — COMMENTED** [review]
This eliminates calls to getExchangeRate and makes conversions to/from LP more standardized.  Big improvement in maintainability.  Appreciate the comments added to tests whose behavior changed.  

Had one concern regarding the `DSTestPlus` change.

File: tests/forge/utils/DSTestPlus.sol (hunk 6 above)
Comment (grandizzy) [reply tests/forge/utils/DSTestPlus.sol]: you're right, not needed, I reverted with https://github.com/ajna-finance/contracts/pull/838/commits/a94c88fe0e6cd6f5e79ab97619ce52db3265bf96

File: src/libraries/external/KickerActions.sol (hunk 3 above)
Comment (grandizzy) [reply src/libraries/external/KickerActions.sol]: removed in https://github.com/ajna-finance/contracts/pull/838/commits/a94c88fe0e6cd6f5e79ab97619ce52db3265bf96

File: tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol (hunk 4 above)
Comment (grandizzy) [reply tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol]: done in https://github.com/ajna-finance/contracts/pull/838/commits/a94c88fe0e6cd6f5e79ab97619ce52db3265bf96

**Review by EdNoepel — APPROVED** [review]
Looks good now; much thanks.

File: src/libraries/internal/Buckets.sol [hunk 7]
```diff
@@ -91,25 +94,41 @@ library Buckets {
...
         uint256 deposit_,
         uint256 collateral_,
-        uint256 bucketPrice_
-    ) internal pure returns (uint256 lp_) {
-        uint256 rate = getExchangeRate(bucketCollateral_, bucketLP_, deposit_, bucketPrice_);
+        uint256 bucketPrice_,
+        Math.Rounding rounding_
+    ) internal pure returns (uint256) {
+        // case when there's no deposit nor collateral in bucket
+        if (deposit_ == 0 && bucketCollateral_ == 0) return Maths.wmul(collateral_, bucketPrice_);
```
Comment (kmbarry1) [inline src/libraries/internal/Buckets.sol:120]: `rounding_` argument not respected here

File: src/libraries/internal/Buckets.sol [hunk 8]
```diff
@@ -91,25 +94,41 @@ library Buckets {
...
+
+        // case when there's deposit or collateral in bucket but no LP to cover
+        if (bucketLP_ == 0) return Maths.wmul(collateral_, bucketPrice_);
```
Comment (kmbarry1) [inline src/libraries/internal/Buckets.sol:123]: `rounding_` argument not respected here

Comment (kmbarry1) [reply src/libraries/internal/Buckets.sol:123]: Could also combine the two conditions into a single case:

```
if ((deposit_ == 0 && bucketCollateral_ == 0) || bucketLP_ == 0) return Maths.wmul(collateral_, bucketPrice_);
```

Comment (mattcushman) [reply src/libraries/internal/Buckets.sol:123]: Comment on the rounding issue: the rounding doesn't matter in the first addition to a bucket (when `bucketLP==0` as this just sets the scale of the bucket LPs, which is arbitrary.

Comment (kmbarry1) [reply src/libraries/internal/Buckets.sol:123]: makes sense

File: src/libraries/internal/Buckets.sol (hunk 7 above)
Comment (mattcushman) [reply src/libraries/internal/Buckets.sol:120]: as noted elsewhere, i think it's ok as first deposit exchange rate is arbitrary

File: tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol [hunk 9]
```diff
@@ -734,15 +736,15 @@ contract ERC20PoolCollateralTest is ERC20HelperContract {
         _assertLenderLpBalance({
             lender:      _lender,
             index:       2570,
-            lpBalance:   6879, // LP should get back to same value as before add / remove collateral
+            lpBalance:   6878, // LP should get back to same value as before add / remove collateral
```
Comment (mattcushman) [inline tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol:739]: note that adding and removing actually consumed one LP due to rounding

**Review by mattcushman — COMMENTED** [review]
lgtm
